lgc/util/ElfNoteEntryInsertionUtil.cpp

```diff
@@ -0,0 +1,279 @@
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+#include "lgc/ElfNoteEntryInsertionUtil.h"
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/Object/ELF.h"
+#include "llvm/Support/Alignment.h"
+#include "llvm/Support/BinaryByteStream.h"
+#include "llvm/Support/BinaryStreamReader.h"
+#include "llvm/Support/BinaryStreamWriter.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace llvm;
+using namespace lgc;
+
+// The implementation of ELF rewriting is based on "Linux Programmer's Manual ELF(5)".
+// In particular, see "Notes (Nhdr)" for the document of the note section.
+
+namespace {
+
+// It is similar to struct NoteHeader defined in llpc/util/vkgcElfReader.h, but
+// it allows us to use Elf_Nhdr_Impl<object::ELF64LE>::Align and it does not
+// have the size limitation of note name.
+using NoteHeader = object::Elf_Nhdr_Impl<object::ELF64LE>;
+
+// An array of zeros. We use it for the note alignment.
+constexpr char ZerosForNoteAlign[NoteHeader::Align] = {'\0'};
+
+// =====================================================================================================================
+// Contents of a section to be shifted and its new offset.
+struct SectionShiftInfo {
+  SmallString<64> section;
+  ELF::Elf64_Off newOffset;
+};
+
+// =====================================================================================================================
+// Generates a StringRef from the given SmallVector.
+//
+// @param input : The SmallVector to be converted to StringRef.
+// @returns     : The generated StringRef.
+template <typename T> StringRef stringRefFromSmallVector(const SmallVectorImpl<T> &input) {
+  return StringRef(reinterpret_cast<const char *>(input.data()), sizeof(T) * input.size());
+}
+
+// =====================================================================================================================
+// Add an entry to the note section.
+//
+// Reference: Linux Programmer's Manual ELF(5) "Notes (Nhdr)".
+//
+// @param noteName              : Name of the note entry.
+// @param noteDesc              : Description of the note entry.
+// @param [out] noteEntryWriter : The .note section where the note entry will be added to.
+void addNoteEntry(StringRef noteName, ArrayRef<uint8_t> noteDesc, BinaryStreamWriter &noteEntryWriter) {
+  NoteHeader noteHeader = {};
+  noteHeader.n_namesz = noteName.size() + 1;
+  noteHeader.n_descsz = noteDesc.size();
+  // TODO: Define a note type to specify cache hash and llpc version. It must
+  // be defined in llvm/include/llvm/BinaryFormat/ELF.h.
+  // Note types with values between 0 and 32 (inclusive) are reserved.
+  noteHeader.n_type = 0;
```

Comment:
Thanks for changing to an LGC API of "add these arbitrary notes". But surely the note type should be passed in the API.